renderdoc/driver/vulkan/vk_pixelhistory.cpp

```diff
@@ -0,0 +1,607 @@
+        m_StencilImageView(stencilImageView),
+        m_ColorImage(colorImage),
+        m_StencilImage(stencilImage),
+        m_PrevState(vk, NULL)
+  {
+    m_pDriver->SetDrawcallCB(this);
+    for(size_t i = 0; i < events.size(); i++)
+      m_Events.insert(std::make_pair(events[i].eventId, events[i]));
+  }
+  ~VulkanPixelHistoryCallback() { m_pDriver->SetDrawcallCB(NULL); }
+  void CopyPixel(uint32_t eid, VkCommandBuffer cmd, uint32_t offset)
+  {
+    CopyPixelParams colourCopyParams = {};
+    colourCopyParams.multisampled = false;    // TODO: multisampled
+    colourCopyParams.srcImage = m_Image;
+    colourCopyParams.srcImageLayout =
+        VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL;    // TODO: image layout
+    colourCopyParams.srcImageFormat = m_Format;
+    colourCopyParams.imageOffset = VkOffset3D{int32_t(m_X), int32_t(m_Y), 0};
+    colourCopyParams.dstBuffer = m_DstBuffer;
+
+    m_pDriver->GetDebugManager()->PixelHistoryCopyPixel(cmd, colourCopyParams, offset);
+
+    const DrawcallDescription *draw = m_pDriver->GetDrawcall(eid);
+    if(draw && draw->depthOut != ResourceId())
+    {
+      // The draw call had a depth image attachment.
+      ResourceId depthImage = m_pDriver->GetResourceManager()->GetLiveID(draw->depthOut);
+      VulkanCreationInfo::Image imginfo = m_pDriver->GetDebugManager()->GetImageInfo(depthImage);
+      CopyPixelParams depthCopyParams = colourCopyParams;
+      depthCopyParams.depthcopy = true;
+      depthCopyParams.srcImage =
+          (VkImage)m_pDriver->GetResourceManager()->GetCurrentResource(depthImage);
+      depthCopyParams.srcImageLayout = VK_IMAGE_LAYOUT_DEPTH_STENCIL_ATTACHMENT_OPTIMAL;
+      depthCopyParams.srcImageFormat = imginfo.format;
+      m_pDriver->GetDebugManager()->PixelHistoryCopyPixel(cmd, depthCopyParams, offset + 16);
+    }
+  }
+
+  void PreDraw(uint32_t eid, VkCommandBuffer cmd)
+  {
+    auto it = m_Events.find(eid);
+    if(it == m_Events.end())
+      return;
+    EventUsage event = it->second;
+    m_PrevState = m_pDriver->GetRenderState();
+
+    m_pDriver->GetRenderState().EndRenderPass(cmd);
+    // Get pre-modification values
+    uint32_t storeOffset = m_EventIndex * sizeof(EventInfo);
+    CopyPixel(eid, cmd, storeOffset);
+
+    // TODO: add all logic for draw calls.
+
+    m_pDriver->GetRenderState() = m_PrevState;
+    if(m_PrevState.graphics.pipeline != ResourceId())
+      m_pDriver->GetRenderState().BeginRenderPassAndApplyState(cmd, VulkanRenderState::BindGraphics);
+  }
+
+  bool PostDraw(uint32_t eid, VkCommandBuffer cmd)
+  {
+    if(m_Events.find(eid) == m_Events.end())
+      return false;
+
+    m_pDriver->GetRenderState().EndRenderPass(cmd);
+
+    uint32_t storeOffset = m_EventIndex * sizeof(EventInfo);
+    CopyPixel(eid, cmd, storeOffset + offsetof(struct EventInfo, postmod));
+
+    m_pDriver->GetRenderState().BeginRenderPassAndApplyState(cmd, VulkanRenderState::BindGraphics);
+
+    // Get post-modification values
+    m_EventIndex++;
```

Comment:
I think this assumes that the callback always calls PreDraw/PostDraw for events in order, which is not necessarily the case. The user can record command buffers A and B in that order, but then submit them in the order B then A. That will cause the callback to hit higher EIDs before lower EIDs.

The solution can either be to get the index in `m_Events` of the EID and use that for the index, or else store the actual order of events seen and use that as a remapping later.

Reply:
Changed.